### DISCUSSION TITLE
Question: Nested query builders

### DISCUSSION THREAD

**Discussion by sowka1995** [react-querybuilder/react-querybuilder#808] [General]
Is it possible to create nested query builders?

I mean situations where we have level 1 fields, but when we select an XYZ field, it has subfields below it.

Example User has the following fields: Id, Name, Email, CreatedBy. If we select CreatedBy, I would like to load the fields available for CreatedBy and display a query builder nested with the fields available for CreatedBy (which is, for example, the User entity).

Maybe something like this can be achieved without nested builders? I tried to create a builder with an empty array of fields and CustomFieldSelector, which has the entire field selection logic, and after selecting it, I tried to dynamically add definitions of such a field to the fields queryBuilder, but unfortunately it doesn't work.

**Comment by jakeboone02** [comment]
Take a look at these (if you haven't already) and let me know if you have any more thoughts. Keep in mind that a query builder is really just a standalone rule group, so in a sense, each subgroup of a query is itself a "nested" query builder. I think the sandbox demonstrates that concept fairly well.

- #515
- #707
- [CodeSandbox example discussed in 707](https://codesandbox.io/p/devbox/nhry9h?file=%2Fsrc%2FApp.tsx)

**Reply by sowka1995** [reply]
The example from discussion 707 is quite nice, but how to add a group with dynamic fields?

What I mean is how to add a different set of fields for "metric" (I want to retrieve them and create them dynamically)

**Reply by sowka1995** [reply]
I think that all I need is to dynamically update the main query builder fields property.
I have a Custom FieldSelector, which is an advanced column selector, and before selecting the values ​​I would like to somehow dynamically insert the field definitions into the query builder.

![image](https://github.com/user-attachments/assets/059e5bfe-8698-43ba-a911-fbe147cf350e)

**Reply by jakeboone02** [reply]
I'm not sure exactly what you mean by "dynamically insert the field definitions". The `fields` prop doesn't have to be static, so you can manage it outside the component and just send a different version in when it changes. Maybe I misunderstood.

**Reply by sowka1995** [reply]
I'm trying to do something like this.
Fields is empty by default. My custom field selector uses an API to retrieve dynamically available fields of various levels. Before calling handleOnChange, I add a new field definition so that QueryBuilder knows such a field and I tried to bypass it with setTimeout, but it doesn't work. It only works after the second click, as if the handleOnChange method was old and remembered the old "fields" state even despite setTImeout

![image](https://github.com/user-attachments/assets/eb4fb8ea-1a07-449d-b71e-ed5fb37cc509)

**Comment by jakeboone02** [comment]
I think this feature, released in v8.8.0, might be what you're looking for: https://react-querybuilder.js.org/docs/tips/subqueries
